File: python/nbody.py
#!/usr/bin/env python3
"""
N-Body Simulation in Python with NumPy
Simulates gravitational interactions between celestial bodies using Newtonian physics.
"""

import numpy as np
import time
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import argparse
import sys


class NBodySimulation:
    def __init__(self, num_bodies=1000, num_iterations=1000, dt=0.01, seed=42):
        """
        Initialize the N-body simulation.

        Args:
            num_bodies: Number of celestial bodies to simulate
            num_iterations: Number of simulation steps
            dt: Time step (delta t)
            seed: Random seed for reproducibility
        """
        self.num_bodies = num_bodies
        self.num_iterations = num_iterations
        self.dt = dt
        self.G = 6.67430e-11  # Gravitational constant

        # Set the random seed for reproducibility
        np.random.seed(seed)

        # Initialize positions, velocities, and masses
        # Using universe size of 1e10 for positions
        self.positions = np.random.uniform(-1e10, 1e10, (num_bodies, 3))
        # Initial velocities near zero
        self.velocities = np.random.uniform(-1e3, 1e3, (num_bodies, 3))
        # Masses between 1e20 and 1e25 kg
        self.masses = np.random.uniform(1e20, 1e25, num_bodies)

        # Initialize forces array
        self.forces = np.zeros((num_bodies, 3))

        # Store position history for visualization
        self.position_history = []

    def compute_forces(self):
        """
        Compute the gravitational forces between all pairs of bodies.
        Optimized to compute each pair only once and apply Newton's third law.

        Returns:
            forces: Array of forces for each body
        """
        # Reset forces array
        self.forces.fill(0.0)

        # Compute pairwise forces (optimized N² algorithm)
        for i in range(self.num_bodies):
            for j in range(i + 1, self.num_bodies):  # Only compute each pair once
                # Vector from body i to body j
                r_vec = self.positions[j] - self.positions[i]
                # Distance squared
                r_squared = np.sum(r_vec * r_vec)
                # Avoid division by zero or extremely small values
                if r_squared < 1e-20:
                    continue

                # Distance
                r = np.sqrt(r_squared)

                # Calculate gravitational force magnitude
                force_mag = self.G * self.masses[i] * self.masses[j] / r_squared

                # Force vector (direction from i to j)
                force_vec = force_mag * r_vec / r

                # Apply Newton's third law: equal and opposite forces
                self.forces[i] += force_vec
                self.forces[j] -= force_vec  # Opposite direction

        return self.forces

    def update_positions_and_velocities(self):
        """
        Update positions and velocities using the computed forces.
        """
        # Calculate accelerations (F = ma, so a = F/m)
        accelerations = self.forces / self.masses[:, np.newaxis]

        # Update velocities (v += a * dt)
        self.velocities += accelerations * self.dt

        # Update positions (p += v * dt)
        self.positions += self.velocities * self.dt

    def run_simulation(self):
        """
        Run the N-body simulation for the specified number of iterations.

        Returns:
            execution_time: Time taken to run the simulation
        """
        # Record the starting time
        start_time = time.time()

        # Store initial positions
        self.position_history.append(self.positions.copy())

        # Run simulation for specified number of iterations
        for _ in range(self.num_iterations):
            self.compute_forces()
            self.update_positions_and_velocities()
            self.position_history.append(self.positions.copy())

        # Calculate execution time
        execution_time = time.time() - start_time
        return execution_time

    def calculate_energy(self):
        """
        Calculate the total energy of the system (kinetic + potential).

        Returns:
            energy: Total energy of the system
        """
        # Initialize energy
        energy = 0.0

        # Calculate kinetic and potential energy
        for i in range(self.num_bodies):
            # Kinetic energy: 0.5 * m * v^2
            velocity_squared = np.sum(self.velocities[i] * self.velocities[i])
            energy += 0.5 * self.masses[i] * velocity_squared

            # Potential energy: -G * m1 * m2 / r
            for j in range(i + 1, self.num_bodies):
                r_vec = self.positions[j] - self.positions[i]
                distance = np.linalg.norm(r_vec)
                if distance > 1e-10:  # Avoid division by zero
                    energy -= self.G * self.masses[i] * self.masses[j] / distance

        return energy

    def visualize(self, num_bodies_to_plot=10, output_file=None):
        """
        Visualize the trajectories of a subset of bodies.

        Args:
            num_bodies_to_plot: Number of bodies to include in the visualization
            output_file: Path to save the visualization. If None, will use a default name.
        """
        fig = plt.figure(figsize=(10, 8))
        ax = fig.add_subplot(111, projection="3d")

        # Select a subset of bodies to plot
        selected_bodies = np.random.choice(
            self.num_bodies, num_bodies_to_plot, replace=False
        )

        for body_idx in selected_bodies:
            # Extract trajectory for this body
            trajectory = np.array(
                [positions[body_idx] for positions in self.position_history]
            )

            # Plot the trajectory
            ax.plot(
                trajectory[:, 0],
                trajectory[:, 1],
                trajectory[:, 2],
                label=f"Body {body_idx}",
            )

            # Mark the final position
            ax.scatter(
                trajectory[-1, 0], trajectory[-1, 1], trajectory[-1, 2], marker="o"
            )

        ax.set_xlabel("X")
        ax.set_ylabel("Y")
        ax.set_zlabel("Z")
        ax.set_title("N-Body Simulation Trajectories")

        # Save the visualization
        if output_file is None:
            output_file = "nbody_python_visualization.png"

        plt.savefig(output_file)
        plt.close()


def parse_args():
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(
        description="N-Body Simulation in Python with NumPy"
    )

    parser.add_argument(
        "--bodies",
        type=int,
        default=1000,
        help="Number of bodies to simulate (default: 1000)",
    )
    parser.add_argument(
        "--iterations",
        type=int,
        default=1000,
        help="Number of simulation iterations (default: 1000)",
    )
    parser.add_argument(
        "--dt", type=float, default=0.01, help="Time step (default: 0.01)"
    )
    parser.add_argument(
        "--seed",
        type=int,
        default=42,
        help="Random seed for reproducibility (default: 42)",
    )
    parser.add_argument(
        "--visualize",
        action="store_true",
        help="Generate visualization of trajectories",
    )
    parser.add_argument(
        "--output", type=str, default=None, help="Output filename for visualization"
    )

    return parser.parse_args()


def main():
    """Main function to run the simulation."""
    # Parse command-line arguments
    args = parse_args()

    # Create and run the simulation
    sim = NBodySimulation(
        num_bodies=args.bodies,
        num_iterations=args.iterations,
        dt=args.dt,
        seed=args.seed,
    )

    # Calculate initial energy
    initial_energy = sim.calculate_energy()
    print(f"Initial system energy: {initial_energy:.6e}")

    # Run the simulation
    execution_time = sim.run_simulation()

    # Calculate final energy
    final_energy = sim.calculate_energy()
    print(f"Final system energy: {final_energy:.6e}")
    print(f"Energy difference: {final_energy - initial_energy:.6e}")

    # Print execution information
    print("Python NumPy Implementation")
    print(f"Number of bodies: {args.bodies}")
    print(f"Number of iterations: {args.iterations}")
    print(f"Time step: {args.dt}")
    print(f"Random seed: {args.seed}")
    print(f"Execution time: {execution_time:.4f} seconds")

    # Visualize if requested
    if args.visualize:
        print("Generating visualization...")
        sim.visualize(num_bodies_to_plot=5, output_file=args.output)
        if args.output:
            print(f"Visualization saved to {args.output}")
        else:
            print("Visualization saved to nbody_python_visualization.png")


if __name__ == "__main__":
    main()

File: julia/nbody.jl
using StaticArrays
using Random, LinearAlgebra, Printf

# Constants
const G = 6.67430e-11  # Gravitational constant

const SV = SVector{3,Float64}

# ------------------------------------------------------
# Body
# ------------------------------------------------------
"""A celestial body with position, velocity, mass and force."""
mutable struct Body
    position::SV # Using SIMD for position (x,y,z)
    velocity::SV # Using SIMD for velocity (vx,vy,vz)
    mass::Float64
    force::SV # Using SIMD for force (fx,fy,fz)
end

Body(position::SV, velocity::SV, mass::Float64) =
    Body(position, velocity, mass, zero(SV))


# ------------------------------------------------------
# NBodySimulation
# ------------------------------------------------------
"""N-body simulation using Newtonian gravitational physics with SIMD optimizations."""
struct NBodySimulation
    bodies::Vector{Body}
    dt::Float64
end

"""Initialize the N-body simulation with random positions, velocities, and masses."""
function NBodySimulation(num_bodies::Int, dt::Float64)
    # Initialize bodies randomly
    Random.seed!(42)

    bodies = Vector{Body}(undef, num_bodies)

    for i in eachindex(bodies)
        # Generate random position
        position = SV(
            randn() * 2e10 - 1e10,  # x
            randn() * 2e10 - 1e10,  # y
            randn() * 2e10 - 1e10,  # z
        )

        # Generate random velocity
        velocity = SV(
            randn() * 2e3 - 1e3,  # vx
            randn() * 2e3 - 1e3,  # vy
            randn() * 2e3 - 1e3,  # vz
        )

        # Generate random mass
        mass = randn() * 1e25 + 1e20

        # Create and add the body
        bodies[i] = Body(position, velocity, mass)
    end

    return NBodySimulation(bodies, dt)
end

"""Compute gravitational forces between all pairs of bodies."""
function compute_forces(nbodysimulation::NBodySimulation)
    bodies = nbodysimulation.bodies
    num_bodies = length(bodies)

    # Reset forces
    for body in bodies
        body.force = zero(SV)
    end

    # Compute pairwise forces (N² algorithm)
    for i in eachindex(bodies)
        body_i = bodies[i]
        for j in (i+1):num_bodies # Optimize: only compute each pair once
            body_j = bodies[j]

            # Vector from body i to body j
            r_vec = body_j.position - body_i.position

            # Distance squared (dot product)
            r_squared = dot(r_vec, r_vec)

            # Avoid division by zero or extremely small values
            r_squared < 1e-20 && continue

            # Distance
            r = sqrt(r_squared)

            # Force magnitude
            force_mag = G * body_i.mass * body_j.mass / r_squared

            # Force vector (direction from i to j, normalized by distance)
            force_vec = r_vec * (force_mag / r)

            # Apply Newton's third law: equal and opposite forces
            body_i.force += force_vec
            body_j.force -= force_vec  # Opposite direction
        end
    end
    return nothing
end

"""Update positions and velocities for all bodies."""
function update_positions_and_velocities(nbodysimulation::NBodySimulation)
    for body in nbodysimulation.bodies
        # Calculate acceleration (F = ma, so a = F/m)
        acceleration = body.force / body.mass

        # Update velocity (v += a * dt)
        body.velocity += acceleration * nbodysimulation.dt

        # Update position (p += v * dt)
        body.position += body.velocity * nbodysimulation.dt
    end
    return nothing
end

"""Run the N-body simulation for the specified number of iterations."""
function run_simulation(nbodysimulation::NBodySimulation, iterations::Int)
    # Record starting time
    start_time = time_ns()

    # Run simulation for specified number of iterations
    for _ in 1:iterations
        compute_forces(nbodysimulation)
        update_positions_and_velocities(nbodysimulation)
    end

    # Calculate execution time in seconds
    end_time = time_ns()

    return (end_time - start_time) * 1e-9  # Time in seconds
end

"""Calculate the total energy of the system (kinetic + potential)."""
function calculate_energy(nbodysimulation::NBodySimulation)
    bodies = nbodysimulation.bodies
    num_bodies = length(bodies)
    energy = 0.0

    # Calculate kinetic energy and potential energy
    for i in eachindex(bodies)
        body_i = bodies[i]

        # Kinetic energy: 0.5 * m * v²
        energy += 0.5 * body_i.mass * dot(body_i.velocity, body_i.velocity)

        # Potential energy: -G * m1 * m2 / r
        for j in (i+1):num_bodies
            body_j = bodies[j]
            diff = body_i.position - body_j.position
            distance = norm(diff)
            energy -= G * body_i.mass * body_j.mass / distance
        end
    end

    return energy
end

# ------------------------------------------------------
# main
# ------------------------------------------------------
function (@main)(args)
    
    println("Starting N-body simulation...")

    # Configuration parameters
    num_bodies = length(args) > 0 ? parse(Int, args[1]) : 2000
    iterations = length(args) > 1 ? parse(Int, args[2]) : 1000
    dt = length(args) > 2 ? parse(Float64, args[3]) : 0.01

    # Create and run the simulation
    simulation = NBodySimulation(num_bodies, dt)

    # Calculate initial energy
    initial_energy = calculate_energy(simulation)
    @printf "Initial system energy: %.4e\n" initial_energy

    # Run simulation and measure time
    execution_time = run_simulation(simulation, iterations)

    # Calculate final energy
    final_energy = calculate_energy(simulation)
    @printf "Final system energy: %.4e\n" final_energy
    @printf "Energy difference: %.4e\n" final_energy - initial_energy

    # Print performance results
    println("Julia Implementation")
    println("Number of bodies: ", num_bodies)
    println("Number of iterations: ", iterations)
    println("Time step: ", dt)
    @printf "Execution time: %.4f seconds\n" execution_time
    println("Done")
end
